src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt

```diff
@@ -195,7 +198,35 @@ class PremisesController(
     bookingId: UUID,
     body: NewExtension
   ): ResponseEntity<Extension> {
-    return super.premisesPremisesIdBookingsBookingIdExtensionsPost(premisesId, bookingId, body)
+    val premises = premisesService.getPremises(premisesId)
+      ?: throw NotFoundProblem(premisesId, "Premises")
+
+    val booking = bookingService.getBooking(bookingId)
+      ?: throw NotFoundProblem(bookingId, "Booking")
+
+    if (booking.premises.id != premises.id) {
+      throw NotFoundProblem(bookingId, "Booking")
+    }
+
+    if (booking.departureDate.isAfter(body.newDepartureDate)) {
+      throw BadRequestProblem(mapOf("newDepartureDate" to "Must be after the Booking's current departure date (${booking.departureDate})"))
+    }
+
+    val extension = bookingService.createExtension(
+      ExtensionEntity(
+        id = UUID.randomUUID(),
+        previousDepartureDate = booking.departureDate,
+        newDepartureDate = body.newDepartureDate,
+        notes = body.notes,
+        booking = booking
+      )
+    )
+
+    booking.departureDate = body.newDepartureDate
+    booking.extensions.add(extension)
+    bookingService.updateBooking(booking)
```

Comment:
Should we do this in a transaction, so if one query fails we don't end up with a weird state where a booking has an extension, but that departure date hasn't been changed or vice versa?

Reply:
Good point, have ammended with a transaction